File: Leetcode/930/930.go
func numSubarraysWithSum(nums []int, goal int) int {
	const MAXN = 6e4 + 5
	pre := make([]int, len(nums)+1)
	for i := 1; i <= len(nums); i++ {
		pre[i] = pre[i-1] + nums[i-1]
	}
	var stats [MAXN][]int
	for i := 0; i < len(nums)+1; i++ {
		stats[pre[i]+goal] = append(stats[pre[i]+goal], i)
	}
	counter := 0
	for r := 0; r < len(nums)+1; r++ {
		if len(stats[pre[r]]) > 0 && r > stats[pre[r]][0] {
			pos := sort.Search(len(stats[pre[r]]), func(i int) bool {
				return stats[pre[r]][i] >= r
			})
			counter += pos
		}
	}
	return counter
}

File: Leetcode/930/930.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        int n = (int)nums.size();
        vector<int> pre(n+1);
        for (int i = 1; i <= n; ++i) {
            pre[i] = pre[i-1] + nums[i-1];
        }
        constexpr int MAXN = 6*1e4+5;
        vector<int> stats[MAXN]{};
        for (int i = 0; i < n+1; ++i) {
            stats[pre[i] + goal].emplace_back(i);
        }
        int counter = 0;
        for (int r = 0; r < n+1; ++r) {
            if (!stats[pre[r]].empty()) {
                if (r <= stats[pre[r]][0]) {
                    continue;
                }
                auto iter = --lower_bound(stats[pre[r]].begin(),stats[pre[r]].end(),r);
                counter += (int)(iter - stats[pre[r]].begin()) + 1;
            }
        }
        return counter;
    }
};